src/triage/component/catwalk/predictors.py

```diff
@@ -159,6 +175,137 @@ def _write_to_db(
+        All entities should have different ranks, so to break ties:
+        - abs_rank uses the 'row_number' function, so ties are broken by the database ordering
+            session.close()
+        - pct_rank uses the output of the abs_rank to compute percentiles
+          (as opposed to raw scores), so it inherits the tie-breaking from abs_rank
+        Args:
+            model_id (int) the id of the model associated with the given predictions
+            matrix_uuid (string) the uuid of the prediction matrix
+        """
+        if not self.save_predictions:
+            logging.info("save_predictions is set to False so there are no predictions to rank")
+            return
+        logging.info(
+            'Beginning ranking of new Predictions for model %s, matrix %s',
+            model_id,
+            matrix_uuid
+        )
+
+        # retrieve a dataframe with only the data we need to rank
+        ranking_df = pandas.DataFrame.pg_copy_from(
+            f"""select entity_id, score, as_of_date, label_value
+            from {matrix_type.string_name}_results.predictions
+            where model_id = {model_id} and matrix_uuid = '{matrix_uuid}'
+            """, engine=self.db_engine)
+
+        sort_seed = None
+        if self.rank_order == 'random':
+            with scoped_session(self.db_engine) as session:
+                sort_seed = session.query(Model).get(model_id).random_seed
+                if not sort_seed:
+                    sort_seed = generate_python_random_seed()
+
+        sorted_predictions, sorted_labels, sorted_arrays = sort_predictions_and_labels(
+            predictions_proba=ranking_df['score'],
+            labels=ranking_df['label_value'],
+            tiebreaker=self.rank_order,
+            sort_seed=sort_seed,
+            parallel_arrays=(ranking_df['entity_id'], ranking_df['as_of_date']),
+        )
+        ranking_df['score'] = sorted_predictions.values
+        ranking_df['as_of_date'] = pandas.to_datetime(sorted_arrays[1].values)
+        ranking_df['label_value'] = sorted_labels.values
+        ranking_df['entity_id'] = sorted_arrays[0].values
+        # at this point, we have the same dataframe that we loaded from postgres,
+        # but sorted based on score and the self.rank_order.
+
+        # Now we can generate ranks using pandas and only using the 'score' column because
+        # our secondary ordering is baked in, enabling the 'first' method to break ties.
+        ranking_df['rank_abs_no_ties'] = ranking_df['score'].rank(ascending=False, method='first')
+        ranking_df['rank_abs_with_ties'] = ranking_df['score'].rank(ascending=False, method='min')
+        ranking_df['rank_pct_no_ties'] = numpy.array([1 - (rank - 1) / len(ranking_df) for rank in ranking_df['rank_abs_no_ties']])
+        ranking_df['rank_pct_with_ties'] = ranking_df['score'].rank(method='min', pct=True)
+
+        # with our rankings computed, upsert these ranks into the predictions table
```

Comment:
just an update, not an upsert, right? That is, it assumes the records are in the predictions table already (at least if I'm reading this correctly)